cmd/aws-janitor-boskos/main.go

```diff
@@ -116,24 +153,32 @@ func cleanResource(res *common.Resource) error {
 	s, err := session.NewSession(aws.NewConfig().WithCredentials(creds))
 	if err != nil {
 		return errors.Wrapf(err, "Failed to create AWS session")
-
 	}
 	logrus.WithField("name", res.Name).Info("beginning cleaning")
 	start := time.Now()
 
+	countSweeps := 0
 	for i := 0; i < *sweepCount; i++ {
 		if err := resources.CleanAll(s, *region, *dryRun); err != nil {
 			if i == *sweepCount-1 {
 				logrus.WithError(err).Warningf("Failed to clean resource %q", res.Name)
+				countSweeps = *sweepCount
+			} else {
+				countSweeps++
 			}
 		}
 		if i < *sweepCount-1 {
 			time.Sleep(sweepSleepDuration)
 		}
 	}
 
-	duration := time.Since(start)
-
-	logrus.WithFields(logrus.Fields{"name": res.Name, "duration": duration.Seconds()}).Info("Finished cleaning")
+	sweepsGauge.WithLabelValues(res.Name).Set(float64(countSweeps))
+	collectMetric(start, res.Name, "clean")
+	logrus.WithFields(logrus.Fields{"name": res.Name, "duration": time.Since(start).Seconds(), "sweeps": countSweeps}).Info("Finished cleaning")
```

Comment:
Looking at this code again, I now realize I was misunderstanding it before - I thought that the specified `sweepCount` was the *maximum* number of attempted sweeps (i.e. retries if there was an error), but actually, I see that it always sweeps that many times, regardless of error or success.

We could keep the metric here, though it's not nearly as meaningful as I thought it was (since it'll basically remain the same for a given configuration). if we keep it, we can just pass `sweepCount` instead of using the separate `countSweeps` variable though.

Reply:
I can drop that for now and in a follow-up do a small refactor in this part to only retry when it failed, and then we add back the metric. WDTY?

Reply:
I think part of the reason it does multiple passes is that the AWS janitor typically does **not** return errors if it fails to delete resources, under the assumption that a later pass might successfully delete the resource (e.g. if there is some lingering dependency that gets cleaned up at a later stage).

This probably isn't ideal, but we'd need to investigate further and fix things to achieve the desired "retry only on error" behavior.

TL;DR we probably want to leave the existing behavior of always sweeping multiple times. The metrics you're adding can help us determine if this is taking too long and we should look into optimizing this process. :)

Reply:
@ixdy sounds good :)

update the code to reflect your feedback, thanks again